cmd/prometheus/main.go

```diff
@@ -611,14 +611,30 @@ func main() {
 		discoveryManagerNotify  discoveryManager
 	)
 
+	// Register the metrics used by both "scrape" and "notify" discovery managers.
+	// The same metrics are used for both discovery managers. Hence the registration
+	// needs to be done here, outside the NewManager() calls, to avoid duplicate
+	// metric registrations.
+	discoveryMetrics, err := discovery.NewMetrics(prometheus.DefaultRegisterer)
```

Comment:
IIUC all these metrics are partitioned by the `name` label. So you _can_ register them separately within each `NewManager` calls if you set the name as `ConstLabels` in the options struct.

You can see this in https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#example-Register and https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#example-GaugeFunc-ConstLabels

Depending on the circumstances, it might be better to use https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#WrapRegistererWith .

Reply:
Thanks a great suggestion! I rebased my branch and pushed a commit to use const labels. I agree that the code looks cleaner now! I tested it by running Prometheus locally again and it worked.